pinot-core/src/main/java/org/apache/pinot/core/data/manager/callback/DataManagerCallback.java

```diff
@@ -0,0 +1,106 @@
+import org.apache.pinot.spi.annotations.InterfaceStability;
+import org.apache.pinot.spi.data.readers.GenericRow;
+import org.apache.pinot.spi.stream.MessageBatch;
+import org.apache.pinot.spi.stream.StreamPartitionMsgOffset;
+
+import java.io.IOException;
+
+/**
+ * Component inject to {@link org.apache.pinot.core.data.manager.SegmentDataManager} for handling extra logic for
+ * upsert-enabled pinot ingestion mode.
+ */
+@InterfaceStability.Evolving
+public interface DataManagerCallback {
+
+  void init() throws IOException;
+
+  /**
+   * create a {@link IndexSegmentCallback} object to allow SegmentDataManager to create proper IndexSegment that supports
+   * either append/upsert mode
+   *
+   * In append-tables callback, this method will create a DefaultIndexSegmentCallback
+   * In upsert-tables callback, this method will create a UpsertDataManagerCallbackImpl
+   *
+   * The {@link IndexSegmentCallback} will be used in the constructor of IndexSegment
+   */
+  IndexSegmentCallback getIndexSegmentCallback();
+
+  /**
+   * process the row after transformation in LLRealtimeSegmentDataManager.processStreamEvents(...) method
+   * it happens after the GenericRow has been transformed by RecordTransformer and before it is indexed by
+   * IndexSegmentImpl, to ensure we can provide other necessary data to the segment metadata.
+   *
+   * In append-tables callback, this method will do nothing
+   * In upsert-tables callback, this method will attach the offset object into the GenericRow object.
+   *
+   * The reason we need this particular logic is that in upsert table, we need to add offset data to the physical data
+   * this will help us to apply the update events from key coordinator to upsert table correctly as the offset
+   * is used as the index to identify which record's virtual column we want to update
+   *
+   * @param row the row of newly ingested and transformed data from upstream
+   * @param offset the offset of this particular pinot record
+   */
+  void processTransformedRow(GenericRow row, StreamPartitionMsgOffset offset);
+
+  /**
+   * process the row after indexing in LLRealtimeSegmentDataManager.processStreamEvents(...) method
+   * it happens after the MutableSegmentImpl has done the indexing of the current row in its physical storage
+   *
+   * In append-tables callback, this method will do nothing
+   * In upsert-tables callback, this method will emit an event to the message queue that will deliver the event to
+   * key coordinator.
+   *
+   * This method ensures that we can emit the metadata for an new entry that pinot just indexed to its internal storage
+   * and let key coordinator to be able to consume those events to process the updates correctly
+   *
+   * @param row the row we just index in the current segment
+   * @param offset the offset associated with the current row
+   */
+  void postIndexProcessing(GenericRow row, StreamPartitionMsgOffset offset);
```

Comment:
rename as `onRealtimeRowIndexed()`. Indicate whether this method should be called in LLC or HLC or both. Indicate whether the method should be called if the row is aggregated (i.e. in the case where `aggregateMetrics` is turned on and the row is not indexed, but aggregated. See https://docs.pinot.apache.org/basics/components/table#tableindexconfig)

Reply:
Updated the documentation indicating this is for LLC and index only (not aggregation)